src/daft-io/src/glob.rs

```diff
@@ -0,0 +1,150 @@
+use std::collections::HashSet;
+
+use lazy_static::lazy_static;
+
+lazy_static! {
+    /// Check if a given char is considered a special glob character
+    /// NOTE: we use the `globset` crate which defines the following glob behavior:
+    /// https://docs.rs/globset/latest/globset/index.html#syntax
+    static ref GLOB_SPECIAL_CHARACTERS: HashSet<char> = {
+        let mut set = HashSet::new();
+        set.insert('*');
+        set.insert('?');
+        set.insert('{');
+        set.insert('[');
+        set
+    };
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct GlobFragment {
+    data: String,
+    first_wildcard_idx: Option<usize>,
+}
+
+impl GlobFragment {
+    pub fn new(data: &str) -> Self {
+        let first_wildcard_idx = if data.is_empty() {
+            None
+        } else if GLOB_SPECIAL_CHARACTERS.contains(&data.chars().nth(0).unwrap()) {
+            Some(0)
+        } else {
+            // Detect any special characters that are not preceded by an escape \
+            let mut idx = None;
+            for (i, window) in data
+                .chars()
+                .collect::<Vec<char>>()
+                .as_slice()
+                .windows(2)
+                .enumerate()
+            {
+                let &[c1, c2] = window else {
+                    unreachable!("Window contains 2 elements")
+                };
+                if (c1 != '\\') && GLOB_SPECIAL_CHARACTERS.contains(&c2) {
+                    idx = Some(i + 1);
+                    break;
+                }
+            }
+            idx
+        };
+        GlobFragment {
+            data: data.to_string(),
+            first_wildcard_idx,
+        }
+    }
+
+    /// Checks if this GlobFragment has any special characters
+    pub fn has_special_character(&self) -> bool {
+        self.first_wildcard_idx.is_some()
+    }
+
+    /// Joins a slice of GlobFragments together with a separator
+    pub fn join(fragments: &[GlobFragment], sep: &str) -> Self {
+        GlobFragment::new(
+            fragments
+                .iter()
+                .map(|frag: &GlobFragment| frag.data.as_str())
+                .collect::<Vec<&str>>()
+                .join(sep)
+                .as_str(),
+        )
+    }
+
+    /// Returns the fragment as a string with the backslash (\) escapes applied
+    /// 1. \\ is cleaned up to just \
+    /// 2. \ followed by anything else is just ignored
+    pub fn escaped_str(&self) -> String {
```

Comment:
look like you're running this every time it is called. It might be more efficient to do it once in `new`